### PR TITLE
[WIP] Multi accounts

### PR DESCRIPTION
**Pull request by jtpedersen**
Issue #383

Support multi accounts from different banks

### CODE REVIEW

File: bank2ynab.conf [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 [DEFAULT]
 # Specify where you save your downloaded CSV files:
-Source Path = %USERPROFILE%\Downloads
+Source Path = /home/jacob/Downloads/
```
Comment (nocalla) [inline bank2ynab.conf:5]: I think we want to keep the generic user here. Not everyone has the same first name.

File: bank2ynab.conf [hunk 2]
```diff
@@ -13,20 +13,21 @@ Source CSV Delimiter = ,
...
+# (see https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes for date format strings)
 Date Format =
 Inflow or Outflow Indicator =
 # Output file formatting
 Output Columns = Date,Payee,Category,Memo,Outflow,Inflow
 Output Filename Prefix = fixed_
 Use Payee for Memo = True
 # Post-processing
-Delete Source File = True
+Delete Source File = False
```
Comment (nocalla) [inline bank2ynab.conf:24]: Why are we changing this default?

File: bank2ynab.conf [hunk 3]
```diff
@@ -13,20 +13,21 @@ Source CSV Delimiter = ,
...
 # Plugins
 Plugin =
 # YNAB API Stuff - don't modify values here, modify in user_configuration.conf
-YNAB API Access Token =
-YNAB Account ID =
+#ffc24c41e613dec1e9eb72bfac15837c7b3014635963f952a01a44d86dac393c
```
Comment (nocalla) [inline bank2ynab.conf:28]: Is this your personal API key??

**Comment by nocalla** [comment]
I'm confused as to what this actually does. The first time we do a write to the API, there's a dialogue asking which account you want to match each bank to, which means you can just dump all your files in one directory and run the script. This seems easier to me - am I missing something?

File: bank2ynab.conf [hunk 4]
```diff
@@ -13,20 +13,21 @@ Source CSV Delimiter = ,
 Header Rows = 1
 Footer Rows = 0
 Input Columns = Date,Payee,Outflow,Inflow,Running Balance
-# (see https://docs.python.org/2/library/datetime.html#id1 for date format strings)
+# (see https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes for date format strings)
```
Comment (nocalla) [inline bank2ynab.conf:16]: This is weird - `develop` already uses this Python 3 documentation link, why is it showing as a change?

File: bank2ynab.conf [hunk 5]
```diff
@@ -340,12 +351,23 @@ Input Columns = skip,Date,skip,skip,Memo,skip,skip,skip,skip,skip,skip,Payee,ski
...
 # Tested Banks: Arbejdernes Landsbank, Merkur Andelskasse, DK Handelsbanken,
 # add more here..
-Source Filename Pattern = Posteringer
+Source Filename Pattern = Al
 Source CSV Delimiter = ;
 Header Rows = 0
 Input Columns = Date,Payee,Inflow,Running Balance
 Date Format = %d-%m-%Y
 
+[DK Portalbank]
```
Comment (nocalla) [inline bank2ynab.conf:360]: This is generating a conflict  - can you remove and pull from `develop` please?

File: bank2ynab/__main__.py [hunk 6]
```diff
@@ -1,22 +1,85 @@
...
+    """
+    root = get_import_root()
+    for budget in budgets:
+        for account in budget.accounts:
+            folder = os.path.join(root, budget.name, account.name)
+            os.makedirs(folder, exist_ok=True)
+            config = os.path.join(folder, CONFIG_FILE)
+            if not os.path.exists(config):
+                with open(config, "w") as f:
+                    json.dump({"bank_id": "DK Bankernes EDB Central"}, f)
```
Comment (nocalla) [inline bank2ynab/__main__.py:34]: Should a specific bank be named here?

File: bank2ynab/__main__.py [hunk 7]
```diff
@@ -1,22 +1,85 @@
...
+
+
+Job = namedtuple("Job", ("budget", "account", "filename", "bank_id"))
+
+
+def move_to_history_folder(job):
```
Comment (nocalla) [inline bank2ynab/__main__.py:40]: I don't see the need for a history folder - is that not the function that YNAB is fulfilling? In any case, I feel like it's a separate feature.

File: bank2ynab/__main__.py [hunk 8]
```diff
@@ -1,22 +1,85 @@
...
+    os.rename(job.filename, os.path.join(folder, name))
+
+
+def process_import_tree(b2y, budgets, api):
+    root = get_import_root()
+    jobs = []
+    for budget in budgets:
+        for account in budget.accounts:
+            account_path = os.path.join(root, budget.name, account.name)
+            bank_id = "DK Bankernes EDB Central"
```
Comment (nocalla) [inline bank2ynab/__main__.py:55]: A specific bank name again.

**Comment by jtpedersen** [comment]
Yes there is a lot of cruft that should be removed and even things that will be fixed by rebase on a recent version of develop. This is why it was marked as WIP "Work In Progress"

I made this to suite my personal workflow and it "works for me" but it does change the general flow of bank2ynab and as such it may not be that you want any of the changes making it into  bank2ynab. Which is why i have not done the necessary cleanup, iff there is any interest in this feature or parts of it, i will clean it up and make a nice PR. But if it is not fitting into the current plan or flow then I can just keep my "weird setup" in my own fork.

File: bank2ynab/__main__.py (hunk 7 above)
Comment (jtpedersen) [reply bank2ynab/__main__.py:40]: I found it to be a nice feature that my csvs does not disapear, upon importing and that i know when they were imported. But yes the data should be in ynab when the script is done.

File: bank2ynab.conf (hunk 5 above)
Comment (jtpedersen) [reply bank2ynab.conf:360]: Ye the whole PR would need a rebase on a recent origin/develop to work. :+1:

File: bank2ynab.conf (hunk 3 above)
Comment (jtpedersen) [reply bank2ynab.conf:28]: it was accidentially commited and revoked some time ago.

Comment (jtpedersen) [reply bank2ynab.conf:28]: https://github.com/bank2ynab/bank2ynab/pull/386#commitcomment-47634433

File: bank2ynab.conf (hunk 1 above)
Comment (jtpedersen) [reply bank2ynab.conf:5]: :+1: then again not everybody uses windows so it should perhaps be some standard pythonpath https://stackoverflow.com/questions/35851281/python-finding-the-users-downloads-folder seems to suggest that it is possible to do.

**Comment by jtpedersen** [comment]
I wrote a more extensive comment on the idea / features on the associated issue #383

File: bank2ynab.conf (hunk 1 above)
Comment (nocalla) [reply bank2ynab.conf:5]: That's what the find_directory function in b2y_utilities does if needed. If Source Path is blank, it uses that.

**Comment by nocalla** [comment]
I like some of the refactoring you've done and it's good to see what doesn't make sense to a fresh pair of eyes, but I think this feature doesn't quite fit with the "one-click" goal of the project since you have to sort into folders beforehand. Thanks for the contribution though.

**Comment by jtpedersen** [comment]
Super. Thank you for taking time to look at it. Please do cherry pick anything you do find useful. Have a nice day.

**Pull request closed without merging**